cve_bin_tool/sources/curl.py

```diff
@@ -0,0 +1,150 @@
+
+from __future__ import annotations
+
+import glob
+import json
+import logging
+import os
+import re
+from typing import Any
+
+import aiohttp
+from bs4 import BeautifulSoup
+
+from cve_bin_tool.async_utils import FileIO, RateLimiter
+from cve_bin_tool.error_handler import (
+    CVEDataForCurlVersionNotInCache,
+    ErrorHandler,
+    ErrorMode,
+)
+from cve_bin_tool.log import LOGGER
+from cve_bin_tool.sources import DISK_LOCATION_BACKUP, DISK_LOCATION_DEFAULT, Source
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+class Curl(Source):
+
+    CACHEDIR = DISK_LOCATION_DEFAULT
+    BACKUPCACHEDIR = DISK_LOCATION_BACKUP
+    LOGGER = LOGGER.getChild("CVEDB")
+    CURL_CVE_FILENAME_TEMPLATE = "curlcve-{}.json"
+
+    def __init__(self, error_mode=ErrorMode.TruncTrace):
+        self.cve_list = None
+        self.cachedir = self.CACHEDIR
+        self.backup_cachedir = self.BACKUPCACHEDIR
+        self.error_mode = error_mode
+        self.session = None
+        self.affected_data = None
+
+    async def get_cve_data(self):
+        await self.fetch_cves()
+        self.get_cve_list()
+
+        return None, self.affected_data
+
+    async def fetch_cves(self):
+        if not self.session:
+            connector = aiohttp.TCPConnector(limit_per_host=19)
+            self.session = RateLimiter(
+                aiohttp.ClientSession(connector=connector, trust_env=True)
+            )
+
+        versions = await self.get_curl_versions(self.session)
+
+        for version in versions:
+            await self.download_curl_version(self.session, version)
+
+        await self.session.close()
+
+    @staticmethod
+    async def get_curl_versions(session: RateLimiter) -> list[str]:
+        regex = re.compile(r"vuln-(\d+.\d+.\d+)\.html")
+        async with await session.get(
+            "https://curl.haxx.se/docs/vulnerabilities.html"
+        ) as response:
+            response.raise_for_status()
+            html = await response.text()
+        matches = regex.finditer(html)
+        return [match.group(1) for match in matches]
+
+    async def download_curl_version(self, session: RateLimiter, version: str) -> None:
+        async with await session.get(
+            f"https://curl.haxx.se/docs/vuln-{version}.html"
+        ) as response:
+            response.raise_for_status()
+            html = await response.text()
+        soup = BeautifulSoup(html, "html.parser")
+        table = soup.find("table")
+        if not table:
+            return
+        headers = table.find_all("th")
+        headers = list(map(lambda x: x.text.strip().lower(), headers))
+        self.LOGGER.debug(headers)
+        rows = table.find_all("tr")
+        json_data = []
+        for row in rows:
+            cols = row.find_all("td")
+            values = (ele.text.strip() for ele in cols)
+            data = dict(zip(headers, values))
+            if data:
+                json_data.append(data)
+        filepath = os.path.abspath(
```

Comment:
Another place for pathlib refactoring. Probably file a separate issue for this file?